lib/ide/ast/impl/src/lib.rs

```diff
@@ -588,158 +582,197 @@ pub enum MacroPatternMatchRaw<T> {
 // === AST ===
 // ===========
 
-// === HasSpan ===
 
-/// Things that can be asked about their span.
-pub trait HasSpan {
-    /// Length of the textual representation of This type in Unicode codepoints.
-    ///
-    /// Usually implemented together with `HasSpan`.For any `T:HasSpan+HasRepr`
-    /// for `t:T` the following must hold: `t.span() == t.repr().len()`.
-    fn span(&self) -> usize;
+// === Tokenizer ===
+
+/// An enum of valid Ast tokens.I
+#[derive(Debug)]
+pub enum Token<'a> { Off(usize), Chr(char), Str(&'a str), Ast(&'a Ast) }
+
+/// Things that can be turned into stream of tokens.
+pub trait HasTokens {
+    /// Feeds TokenBuilder with stream of tokens obtained from `self`.
+    fn feed(&self, consumer:&mut impl TokenConsumer);
 }
 
-/// Counts codepoints.
-impl HasSpan for char {
-    fn span(&self) -> usize {
-        1
-    }
+/// Helper trait for Tokenizer, which consumes the token stream.
+pub trait TokenConsumer {
+    /// consumes one token
+    fn feed(&mut self, val:Token);
 }
 
-/// Counts codepoints.
-impl HasSpan for String {
-    fn span(&self) -> usize {
-        self.as_str().span()
+
+impl HasTokens for &str {
+    fn feed(&self, consumer:&mut impl TokenConsumer) {
+        consumer.feed(Token::Str(self));
```

Comment:
why not `consumer.feed(self.into())` ? There should be conversion instance defined

Reply:
I have been told that on rust channel that implementing `From` for references isn't normal.

Reply:
I don't think so. We have a lot of such from s implemented.

Reply:
Nevertheless, we don't use Token anywhere but here, so implementing `From` would just add 30LOC without any real benefit.